# 7. 결과부족이 발생할 경우 null과 Failure를 사용하라

## 함수가 원하는 결과를 만들어낼 수 없을 때
- Server로부터 데이터를 읽으려했으나, 네트워크 문제로 읽어오지 못한 경우
- 조건에 맞는 첫번째 요소를 찾지못했을 경우
- Text를 파싱하여 객체를 만드는데, 알맞은 형식이 아닌경우
- …

## null 또는 Failure를 통해서 위 문제를 해결하자
- 예측할 수 있는 경우에는 null이나 sealed class를 사용하자.
    - 추가적인 정보를 전달할 일이 없다면 null을 리턴하자.
    - 추가적인 정보를 포함해야 한다면 sealed class를 리턴하자.
- 예측 할 수 없는 경우에 Exception을 사용하자.

### 선택 할 수 있는 방법
- null(없음)
    - ?. (safeCall), ?: (Elvis Operator)를 통해서 안전한 null연산 수행이 가능하다.
- Sealed Class (보통 Failure라는 이름을 붙인다)을 리턴
    ```kotlin
    sealed class Result<out T>
    class Success<out T>(val result:T): Result<T>
    class Failure(val throwable:Throwable): Result<Nothing>
    ```
    - Result 타입을 when으로 받아서 적절한 처리를 하는 것도 가능하다.
- Exception Throw
    - 정보를 전달하는 방법으로 사용할 수 없다.
    - 예외적인 상황일 때만 발생시키는게 좋다.
    - try-catch 사용시 컴파일러 최적화에 문제가 생길 수 있다.
    - 예외적인 상황에서만 발생하기 때문에 Test복잡성이 증가한다.
    - UnCheckedException(예외처리를 강제하지않음)이기 때문에 문서화가 잘되어있지 않다면 어려움을 겪을 수 있다.